src/vdom.rs

```diff
@@ -423,49 +412,59 @@ impl VdomWeak {
     /// `render` instead.
     pub fn schedule_render(&self) {
         debug!("VdomWeak::schedule_render");
-        wasm_bindgen_futures::spawn_local(self.render().map_err(|_| ()));
+
+        let future = self.render();
+
+        wasm_bindgen_futures::spawn_local(async move {
+            let _ = future.await;
+        });
     }
 
     /// Schedule a render to occur during the next animation frame and return a
     /// future that will complete once the render has finished.
     ///
     /// If you don't want to do more things after the render completes, then use
     /// `schedule_render` instead of `render`.
-    pub fn render(&self) -> impl Future<Item = (), Error = VdomDroppedError> {
-        futures::future::ok(self.inner.upgrade())
-            .and_then(|inner| inner.ok_or(()))
-            .map_err(|_| VdomDroppedError {})
-            .and_then(|inner| {
-                let promise = inner.shared.render_scheduled.take().unwrap_or_else(|| {
-                    js_sys::Promise::new(&mut |resolve, reject| {
-                        let vdom = VdomWeak {
-                            inner: Rc::downgrade(&inner),
-                        };
-                        with_animation_frame(move || match vdom.inner.upgrade() {
-                            None => {
-                                warn!("VdomWeak::render: vdom unmounted before we could render");
-                                let r = reject.call0(&JsValue::null());
-                                debug_assert!(r.is_ok());
-                            }
-                            Some(inner) => {
-                                let mut exclusive = inner.exclusive.borrow_mut();
-                                exclusive.render();
-
-                                // We did the render, so take the promise away
-                                // and let future `render` calls request new
-                                // animation frames.
-                                let _ = inner.shared.render_scheduled.take();
-
-                                let r = resolve.call0(&JsValue::null());
-                                debug_assert!(r.is_ok());
-                            }
-                        });
-                    })
-                });
-                inner.shared.render_scheduled.set(Some(promise.clone()));
-                JsFuture::from(promise)
-                    .map(|_| ())
-                    .map_err(|_| VdomDroppedError {})
-            })
+    pub fn render(&self) -> impl Future<Output = Result<(), VdomDroppedError>> {
+        let inner = self.inner.upgrade();
+
+        async move {
```

Comment:
This was the easiest way to deal with a lifetime issue that I wasn't acquainted with before